website/integrations: Add instructions for Mattermost Team Edition (#14802)
goauthentik/authentik: website/integrations/services/mattermost-team-edition/index.md

> ---

Reviewer: This page should be added to the [integration sidebar](https://github.com/goauthentik/authentik/blob/main/website/sidebars/integrations.mjs).

title: Integrate with Mattermost Team Edition
sidebar_label: Mattermost Team Edition
support_level: community
---

## What is Mattermost Team Edition

> Mattermost is an open source, real-time collaboration platform. It provides chat, audio/video calling, screen sharing, and a plugin architecture for extending its capabilities. Mattermost Team Edition is the free, open-source version of the product.
>
> -- https://mattermost.com/

> Mattermost Team Edition does not natively support generic single sign-on. With a manual edit to your configuration, you can use Mattermost's GitLab integration with authentik's OAuth2/OpenID Provider to authenticate using authentik. When this configuration is complete, Mattermost will display a login button that uses the GitLab icon on it, but it will be using your authentik system for authentication. You do not need to have GitLab installed or available. This setup does not use GitLab in any way.

Reviewer: ```suggestion
:::info
Mattermost Team Edition does not natively support generic single sign-on. However, with direct access to the server’s filesystem, you can manually configure Mattermost to use its GitLab integration for authentication via authentik’s OAuth2/OpenID Provider. This requires editing the config.json file directly, as the necessary settings are not available through the web interface. If you are using a hosted version of Mattermost without filesystem access, you will not be able to complete this setup. Once configured, Mattermost will display a login button with the GitLab icon, but authentication will be handled entirely by authentik - GitLab itself is not required or used in any way.
:::
```

## Preparation

The following placeholders are used in this guide:

- `mattermost.company` is the FQDN of the Mattermost installation.
- `authentik.company` is the FQDN of the authentik installation.

> You will need direct access to the filesystem for your Mattermost Team Edition server. The configurations that are required can only be set by editing the `config.json` file directly. If you use a hosted version of Mattermost where you cannot access the files, you will not be able to complete these instructions.

Reviewer: ```suggestion
```

Moved to ^^.

:::note
This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

## authentik configuration

> To support the integration of Mattermost Team Edition with authentik, you need to create an application/provider pair in authentik.

Reviewer: ```suggestion
To support the integration of Mattermost Team Edition with authentik, you need to create property mappings and an application/provider pair in authentik.
```

### Create property mappings

> You need to create two scope mappings.

Reviewer: ```suggestion
```

1. Log in to authentik as an administrator and open the authentik Admin interface.
2. Navigate to **Customization** > **Property Mappings** and click **Create**. Create a **Scope Mapping** with the following settings:
    - **Name**: Set an appropriate name.
    - **Scope Name**: `id`
    - **Description**: Set an appropriate description, if desired.
    - **Expression**:
        ```python
        return {

>             "id": request.user.attributes.get("mattermostId", request.user.id)

Reviewer 2: ```suggestion
            "id": request.user.attributes.get("mattermostId", request.user.pk),
```

        }
        ```
3. Click **Create** again, and create a **Scope Mapping** with the following settings:
    - **Name**: Set an appropriate name.
    - **Scope Name**: `username`
    - **Description**: Set an appropriate description, if desired.
    - **Expression**:
        ```python
        return {

>             "username": request.user.username

Reviewer 2: ```suggestion
            "username": request.user.username,
```

        }
        ```

:::note

> The `mattermostId` property mapping is optional. Without it, Mattermost will use a variation on users' email addresses as their ID on the system. This leads to userids that look like **person-example.com** for a user whose email address is `person@example.com`. Since these are the nicknames in a chat system, that may be undesirable. Using the `mattermostId` property you can select or construct a different identifier that might be more aesthetically pleasing.

Reviewer: ```suggestion
The `id` property mapping is optional. If omitted, Mattermost will generate user IDs based on email addresses, resulting in names such as `person-example.com` for `person@example.com`. Since these IDs serve as nicknames, this format may be undesirable.
```

Not sure I love the exact phrasing here (or my suggestion), but I do like this part from the MR: *"Using the `mattermostId` property you can select or construct a different identifier that might be more aesthetically pleasing."* You could potentially revise the expression in step 2 to suggest additional formats. However, according to your comment, this ID functions as a "nickname", wouldn’t `request.user.username` be more appropriate?

:::

### Create an application and provider in authentik

1. Log in to authentik as an administrator and open the authentik Admin interface.
2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

> - **Application**: provide a descriptive name, an optional group for the type of application, and the policy engine mode

Reviewer: ```suggestion
- **Application**: provide a descriptive name, an optional group for the type of application, and the policy engine mode.
```

>     - Under UI Settings set the Launch URL to <kbd>https://<em>mattermost.company</em>/oauth/gitlab/login</kbd>

Reviewer: We don't usually include the launch url in our docs as it is constructed automatically. However, if it needs to be this URL, we can keep it.

- **Choose a Provider type**: select **OAuth2/OpenID Connect** as the provider type.
- **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
    - Note the **Client ID**,**Client Secret**, and **slug** values because they will be required later.

>     - Set a `Strict` redirect URI to <kbd>https://<em>mattermost.company</em>/signup/gitlab/complete</kbd>

Reviewer: ```suggestion
    - Set a `Strict` redirect URI to `https://mattermost.company/signup/gitlab/complete`.
```

We don't em/kbd URLs anymore, you can just use inline codeblock and hope the user read the preparations section. (I think the style guide was updated, I'd have to check.)

    - Select any available signing key.
    - Under **Advanced protocol settings**, add the scopes you just created to the list of selected scopes.
- **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

## Mattermost Team Edition configuration

> 1. Find the `config.json` file for your Mattermost installation ([documentation](https://docs.mattermost.com/configure/configuration-settings.html)). Open it in a text editor.

Reviewer: Could we remove the anchor, "Open it in a text editor",  and mabye merge steps 1 + 2? I'd probably do something similar to `To support the integration of Mattermost Team Editor with authentik, you'll need to update the `config.json` file of your Mattermost deployment:`

> 2. Change the `GitLabSettings` section of the file to look like the following. Insert your **client secret**, **client id**, and **slug** in the appropriate places.

Reviewer: This would be removed with the comment for line 81. "Insert x, y, and z" is kind of implied due to the used placeholders and the section to change is made evident by the indenting, format, and whatnot.

> ```json

Reviewer: ```suggestion
```json showLineNumbers title="/opt/mattermost/config/config.jsonl"
```

"GitLabSettings": {
    "Enable": true,

>     "Secret": "**client secret**",

Reviewer: ```suggestion
    "Secret": "<client_secret>",
```

>     "Id": "**client id**",

Reviewer: ```suggestion
    "Id": "<client_id>",
```

    "Scope": "",
    "AuthEndpoint": "https://authentik.company/application/o/authorize/",
    "TokenEndpoint": "https://authentik.company/application/o/token/",
    "UserAPIEndpoint": "https://authentik.company/application/o/userinfo/",

>     "DiscoveryEndpoint": "https://authentik.company/application/o/**slug**/.well-known/openid-configuration",

Reviewer: ```suggestion
    "DiscoveryEndpoint": "https://authentik.company/application/o/<application_slug>/.well-known/openid-configuration",
```

>     "ButtonText": "Login with authentik",

Reviewer: ```suggestion
    "ButtonText": "Log in with authentik",
```

    "ButtonColor": "#000000"
},
```

3. On the Mattermost **Authentication** > **Signup** options (`https://mattermost.company/admin_console/authentication/signup`) make sure that **Enable Account Creation** is **true**.
4. After saving the changes to the `config.json` file, restart Mattermost.

> ## Additional Resources

Reviewer: Please add a configuration verification section right above this section
